### PR TITLE
Introducing unit tests.

### PR DESCRIPTION
**Pull request by ewastempel**
This PR is to enable API mocking for use in unit tests and to provide an example on how to implement such unit tests.

The above has been achieved through:
1. Introducing aws client's interfaces so that it is possible to mock their APIs (in main.go)
2. Passing client's interfaces into functions rather than defining them inside the functions.
3. Introducing simple unit tests for getSecret and getParameter functions in main_test.go 

More unit tests and more code refactoring will be needed in the future PRs for better code coverage. This is just a mechanism and an example of how it can be achieved.

### CODE REVIEW

File: instance-scheduler/main.go [hunk 1]
```diff
@@ -38,38 +38,47 @@ aws ssm get-parameter --name environment_management_arn --with-decryption --prof
...
+	})
+
 	if err != nil {
 		log.Fatal(err)
 	}
 	return *result.SecretString
 }
 
-func getNonProductionAccounts(cfg aws.Config, skipAccountNames string) map[string]string {
+func getNonProductionAccounts(secretId string, environments string, skipAccountNames string) map[string]string {
```
Comment (julialawrence) [inline instance-scheduler/main.go:80]: It doesn't look like secretId is being used in this function. Is it for future-proofing?

File: instance-scheduler/main.go [hunk 2]
```diff
@@ -268,7 +277,15 @@ func handler(ctx context.Context, request InstanceSchedulingRequest) (events.API
 		log.Fatal(err)
 	}
 
-	accounts := getNonProductionAccounts(cfg, skipAccounts)
+	// Get the secret ID (ARN) of the environment management secret from the parameter store
+	ssmClient := ssm.NewFromConfig(cfg)
```
Comment (julialawrence) [inline instance-scheduler/main.go:281]: Is moving the retrieval of the param and secret out of the getNonProdAccounts function make it easier to mock/test?

**Review by julialawrence — COMMENTED** [review]
I think I understand what this is doing. It looks good. I have a couple of questions/comments.

File: instance-scheduler/main.go (hunk 1 above)
Comment (ewastempel) [reply instance-scheduler/main.go:80]: good catch, going to fix it

File: instance-scheduler/main.go (hunk 2 above)
Comment (ewastempel) [reply instance-scheduler/main.go:281]: Yes. It's a bit of a domino effect. It is impossible to test functions if you define clients inside them (as you cannot mock them), hence getSecret and getParameter now accept clients. Now, I could pass clients into the getNonProdAccounts accounts and then define/assign secretId and environments in there, but that means in order to test getNonProdAccounts, you would need to mock two objects, instead you do not need to mock anything, so less work.

Comment (julialawrence) [reply instance-scheduler/main.go:281]: Makes sense.

**Review by julialawrence — APPROVED** [review]
LGTM

**Review by gfou-al — APPROVED** [review]
Looks good

**Pull request merged**